python/pyarrow/_fs.pyx

```diff
@@ -0,0 +1,515 @@
+            )
+        return self.stats.size()
+
+    @property
+    def extension(self):
+        """The file extension"""
+        return frombytes(self.stats.extension())
+
+    @property
+    def mtime(self):
+        """The time of last modification, if available.
+
+        Returns
+        -------
+        mtime : datetime.datetime
+        """
+        cdef PyObject *out
+        check_status(PyDateTime_from_TimePoint(self.stats.mtime(), &out))
+        return PyObject_to_object(out)
+
+
+cdef class Selector:
+    """File and directory selector.
+
+    It containt a set of options that describes how to search for files and
+    directories.
+
+    Parameters
+    ----------
+    base_dir : Union[str, pathlib.Path]
+        The directory in which to select files.
+        If the path exists but doesn't point to a directory, this should be an
+        error.
+    allow_non_existent : bool, default False
+        The behavior if `base_dir` doesn't exist in the filesystem.
+        If false, an error is returned.
+        If true, an empty selection is returned.
+    recursive : bool, default False
+        Whether to recurse into subdirectories.
+    """
+    cdef CSelector selector
+
+    def __init__(self, base_dir='', bint allow_non_existent=False,
+                 bint recursive=False):
+        self.base_dir = base_dir
+        self.recursive = recursive
+        self.allow_non_existent = allow_non_existent
+
+    @property
+    def base_dir(self):
+        return pathlib.Path(self.selector.base_dir)
+
+    @base_dir.setter
+    def base_dir(self, base_dir):
+        self.selector.base_dir = _path_to_bytes(base_dir)
+
+    @property
+    def allow_non_existent(self):
+        return self.selector.allow_non_existent
+
+    @allow_non_existent.setter
+    def allow_non_existent(self, bint allow_non_existent):
+        self.selector.allow_non_existent = allow_non_existent
+
+    @property
+    def recursive(self):
+        return self.selector.recursive
+
+    @recursive.setter
+    def recursive(self, bint recursive):
+        self.selector.recursive = recursive
+
+
+cdef class FileSystem:
+    """Abstract file system API"""
+
+    cdef:
+        shared_ptr[CFileSystem] wrapped
+        CFileSystem* fs
+
+    def __init__(self):
+        raise TypeError('dont initialize me')
+
+    cdef init(self, const shared_ptr[CFileSystem]& wrapped):
+        self.wrapped = wrapped
+        self.fs = wrapped.get()
+
+    def get_target_stats(self, paths_or_selector):
+        """Get statistics for the given target.
+
+        Any symlink is automatically dereferenced, recursively. A non-existing
+        or unreachable file returns an Ok status and has a FileType of value
+        NonExistent. An error status indicates a truly exceptional condition
+        (low-level I/O error, etc.).
+
+        Parameters
+        ----------
+        paths_or_selector: Union[Selector, List[Union[str, pathlib.Path]]]
```

Comment:
We don't really use this syntax when documenting types, but a more literary form (e.g. "Selector or list of paths").
Also, the C++ API allows a single path here.

Reply:
Then the output type would change to a single FileStats object instead of a list of FileStat objects. How about a different method for a single input?

Reply:
I see. I can't think of a good for another method. Let's keep it like this?

Reply:
Me neither :), only the singular and plural forms, but that is error-prone.